daemon/cmd/ipcache.go

```diff
@@ -87,3 +104,126 @@ func containsSubnet(outer, inner net.IPNet) bool {
 
 	return outerBits == innerBits && outerOnes <= innerOnes && outer.Contains(inner.IP)
 }
+
+// restoreIPCache dumps the existing (old) bpf ipcache, adding relevant information
+// back in to the ipcache / identity allocator.
+//
+// For all local (cidr / remote-node) identities, it will add a stub metadata entry
+// that ensures the prefix exists and it gets the same numeric identity before. It
+// reserves those local numeric IDs from allocation to ensure they are available.
+//
+// For ingress IPs, it will add those to the ipcache and configure the local node
+// accordingly.
+//
+// This *must* be called before initMaps(), which will clobber the "old" ipcache.
+func (d *Daemon) restoreIPCache() error {
+	ingressIPs := make([]netip.Prefix, 0, 2)
+	// need to preserve this so we can remove it later.
+	d.restoredCIDRs = map[netip.Prefix]identity.NumericIdentity{}
+
+	// Dump the bpf ipcache, recording any prefixes with local or ingress
+	// numeric identities.
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+
+		if isLocalIdentity(nid) {
+			d.restoredCIDRs[k.Prefix()] = nid
+		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
+			ingressIPs = append(ingressIPs, k.Prefix())
+		}
+	})
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error dumping ipcache: %w", err)
+	}
+	// DumpWithCallback() leaves the ipcache map open, must close before opened for
+	// parallel mode in Daemon.initMaps()
+	ipcachemap.IPCacheMap().Close()
+
+	// Now that the map is dumped,
+	// - upsert relevant metadata in to the ipcache
+	// - reserve all existing CIDR identities
+	// - add Ingress IPs to local Node.
+	metaUpdates := make([]ipcache.MU, 0, len(ingressIPs)+len(d.restoredCIDRs))
+	nidsToReserve := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+
+	for prefix, nid := range d.restoredCIDRs {
+		nidsToReserve = append(nidsToReserve, nid)
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
+		})
+	}
+	for _, prefix := range ingressIPs {
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: ipcachetypes.NewResourceID(ipcachetypes.ResourceKindDaemon, "", "ingress"),
+			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+		})
+
+		// Set any restored ingress IPs back on the LocalNode object
+		d.nodeLocalStore.Update(func(n *node.LocalNode) {
```

Comment:
Mildly curious, I don't see this logic in the current tree. Is this a bug? Why is this logic now needed? Or did it exist somewhere else and I missed it?

Reply:
It was [here](https://github.com/cilium/cilium/pull/28673/files#diff-2ed5415f522d0b72066d62527fb40dbea9bf745f45d5e2d47f7715680f165fcaL543). AIUI we're trying to migrate off of the top-level local node mutators, so I did that.

Reply:
+1 on the cleanup.

(I'm not a fan of how we ingress IP restoration is cluttered all over the place (i.e. the IP stored here still needs to be allocated from IPAM later on), but alas, that's a refactor for another time)

Reply:
Oh I see, it's just directly calling the Update function openly instead of using the helper. There's no functional change here right?

Reply:
Correct, behavior is identical. The helper can't go away yet, but this removes one more user.